File: packages/telemetry/typescript/examples/test_capture_nesting.ts
/**
 * Nested capture() context merging — Latitude telemetry example.
 *
 * Verifies that nested capture() calls correctly merge context onto the spans
 * created within each nesting level:
 * - tags: merge and deduplicate
 * - metadata: shallow merge (child overrides parent for same keys)
 * - sessionId/userId: last-write-wins (innermost capture wins)
 *
 * Nested captures reuse the active trace/root span and only MERGE context — they
 * don't create a span per level — so each LLM span reflects the merged context
 * of the capture it ran inside.
 *
 * Required env vars:
 * - LATITUDE_API_KEY
 * - LATITUDE_PROJECT_SLUG
 * - OPENAI_API_KEY
 *
 * Install: npm install openai
 */

import { randomUUID } from "node:crypto"
import OpenAI from "openai"
import { capture, Latitude } from "../src"

const latitude = new Latitude({
  apiKey: process.env.LATITUDE_API_KEY!,
  project: process.env.LATITUDE_PROJECT_SLUG!,
  instrumentations: { openai: OpenAI },
  disableBatch: true,
})

const openai = new OpenAI()

const MODEL = "gpt-5.5"
// gpt-5.5 is a reasoning model: budget must cover reasoning + the visible answer (else finish_reason "length").
const MAX_TOKENS = 2000
const SYSTEM = "You are a helpful assistant participating in a telemetry QA test. Keep answers concise."
// Shared run id so the whole run is findable via a single tag, while the per-capture
// sessionId/userId/metadata stay as the merge-semantics fixture they exist to test.
const RUN = `capture-nesting-${randomUUID().slice(0, 8)}`

async function say(prompt: string) {
  const response = await openai.chat.completions.create({
    model: MODEL,
    messages: [
      { role: "system", content: SYSTEM },
      { role: "user", content: prompt },
    ],
    max_completion_tokens: MAX_TOKENS,
  })
  return response.choices[0]?.message?.content
}

async function toolConversation() {
  const tools = [
    {
      type: "function" as const,
      function: {
        name: "get_weather",
        description: "Get the current weather for a city",
        parameters: {
          type: "object",
          properties: { city: { type: "string" } },
          required: ["city"],
        },
      },
    },
  ]
  const messages: OpenAI.Chat.ChatCompletionMessageParam[] = [
    { role: "system", content: SYSTEM },
    { role: "user", content: "What's the weather in San Francisco? Use get_weather, then answer in one short sentence." },
  ]

  const first = await openai.chat.completions.create({ model: MODEL, messages, tools, max_completion_tokens: MAX_TOKENS })
  const toolCall = first.choices[0]?.message?.tool_calls?.[0]
  messages.push(first.choices[0]!.message)
  messages.push({
    role: "tool",
    tool_call_id: toolCall!.id,
    content: JSON.stringify({ city: "San Francisco", temperatureC: 21, conditions: "sunny" }),
  })

  const second = await openai.chat.completions.create({ model: MODEL, messages, tools, max_completion_tokens: MAX_TOKENS })
  return second.choices[0]?.message?.content
}

async function decoratorStyleNesting() {
  // Outer capture; the first LLM span carries the outer context verbatim.
  return capture(
    "outer-capture",
    async () => {
      const outerResponse = await say("Say 'First call' in exactly 2 words.")

      // Inner capture runs the tool conversation so we verify the MERGED context
      // (tags deduped, sessionId/userId overridden, metadata shallow-merged)
      // lands on the tool-conversation spans.
      const innerResponse = await capture("inner-capture", toolConversation, {
        tags: ["inner-tag", "shared-tag", RUN, "capture-nesting-ts"], // shared-tag deduped
        sessionId: "inner-session", // overrides outer-session
        userId: "inner-user", // overrides outer-user
        metadata: { inner_key: "inner_value", shared_key: "inner_shared" }, // shared_key overridden
      })

      return { outerResponse, innerResponse }
    },
    {
      tags: ["outer-tag", "shared-tag", RUN, "capture-nesting-ts"],
      sessionId: "outer-session",
      userId: "outer-user",
      metadata: { outer_key: "outer_value", shared_key: "outer_shared" },
    },
  )
}

async function deeplyNested() {
  return capture(
    "level-1",
    async () => {
      const nested = await capture(
        "level-2",
        async () => {
          const level3Result = await capture("level-3", () => say("Say 'Level 3' in exactly 2 words."), {
            tags: ["level-3-tag", RUN, "capture-nesting-ts"],
            sessionId: "level-3-session",
            userId: "level-3-user",
            metadata: { level: 3, shared: "level-3" },
          })
          const level2Result = await say("Say 'Level 2' in exactly 2 words.")
          return { level3Result, level2Result }
        },
        {
          tags: ["level-2-tag", RUN, "capture-nesting-ts"],
          sessionId: "level-2-session",
          userId: "level-2-user",
          metadata: { level: 2, shared: "level-2" },
        },
      )
      const level1Result = await say("Say 'Level 1' in exactly 2 words.")
      return { nested, level1Result }
    },
    {
      tags: ["level-1-tag", RUN, "capture-nesting-ts"],
      sessionId: "level-1-session",
      userId: "level-1-user",
      metadata: { level: 1, shared: "level-1" },
    },
  )
}

async function main() {
  await latitude.ready

  console.log(`Run tag: ${RUN}`)

  console.log("\n1. Testing nested captures (outer → inner tool conversation)...")
  console.log("Result:", await decoratorStyleNesting())

  console.log("\n2. Testing deeply nested captures (3 levels)...")
  console.log("Result:", await deeplyNested())

  await latitude.flush()
  await latitude.shutdown()

  console.log("\nExpected merge on the inner-capture (tool-conversation) spans:")
  console.log("- tags = [outer-tag, shared-tag, inner-tag] (deduplicated) + run/lang tags")
  console.log("- sessionId = inner-session, userId = inner-user (overridden)")
  console.log("- metadata.shared_key = inner_shared (overridden); outer_key + inner_key both present")
}

main().catch(console.error)

File: packages/telemetry/python/examples/test_capture_nesting.py
"""
Test nested capture() context merging.

This verifies that nested capture() calls correctly merge context:
- tags: merge and deduplicate
- metadata: shallow merge (child overrides parent for same keys)
- session_id/user_id: last-write-wins

Required env vars:
- LATITUDE_API_KEY
- LATITUDE_PROJECT_SLUG
- OPENAI_API_KEY

Install: uv add openai
"""

import os

from openai import OpenAI

from latitude_telemetry import capture, init_latitude

# Initialize telemetry pointing to local instance
latitude = init_latitude(
    api_key=os.environ["LATITUDE_API_KEY"],
    project_slug=os.environ["LATITUDE_PROJECT_SLUG"],
    instrumentations=["openai"],
    disable_batch=True,
)


@capture(
    "outer-capture",
    {
        "tags": ["outer-tag", "shared-tag"],
        "session_id": "outer-session",
        "user_id": "outer-user",
        "metadata": {"outer_key": "outer_value", "shared_key": "outer_shared"},
    },
)
def outer_function():
    """Outer capture with initial context."""
    client = OpenAI()

    # First LLM call in outer context
    response1 = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[{"role": "user", "content": "Say 'First call' in exactly 2 words."}],
        max_tokens=50,
    )

    # Call inner function with nested capture
    inner_result = inner_function()

    # Second LLM call (should still have outer context)
    response2 = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[{"role": "user", "content": "Say 'Second call' in exactly 2 words."}],
        max_tokens=50,
    )

    return {
        "outer_response": response1.choices[0].message.content,
        "inner_response": inner_result,
        "second_outer_response": response2.choices[0].message.content,
    }


@capture(
    "inner-capture",
    {
        "tags": ["inner-tag", "shared-tag"],  # shared-tag should be deduplicated
        "session_id": "inner-session",  # should override outer-session
        "user_id": "inner-user",  # should override outer-user
        "metadata": {"inner_key": "inner_value", "shared_key": "inner_shared"},  # shared_key should be overridden
    },
)
def inner_function():
    """Inner capture with context that should merge with outer."""
    client = OpenAI()

    response = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[{"role": "user", "content": "Say 'Inner call' in exactly 2 words."}],
        max_tokens=50,
    )

    return response.choices[0].message.content


def test_nested_capture_with_callback():
    """Test nested capture using callback pattern."""

    def outer_callback():
        client = OpenAI()

        def inner_callback():
            inner_client = OpenAI()
            response = inner_client.chat.completions.create(
                model="gpt-4o-mini",
                messages=[{"role": "user", "content": "Say 'Callback inner' in exactly 2 words."}],
                max_tokens=50,
            )
            return response.choices[0].message.content

        # Inner capture with callback pattern
        inner_result = capture(
            "callback-inner",
            inner_callback,
            {
                "tags": ["callback-inner"],
                "session_id": "callback-inner-session",
                "user_id": "callback-inner-user",
                "metadata": {"callback_inner": "value"},
            },
        )

        response = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": "Say 'Callback outer' in exactly 2 words."}],
            max_tokens=50,
        )

        return {
            "inner": inner_result,
            "outer": response.choices[0].message.content,
        }

    return capture(
        "callback-outer",
        outer_callback,
        {
            "tags": ["callback-outer"],
            "session_id": "callback-outer-session",
            "user_id": "callback-outer-user",
            "metadata": {"callback_outer": "value"},
        },
    )


if __name__ == "__main__":
    print("Testing nested capture with decorators...")
    result1 = outer_function()
    print(f"Decorator result: {result1}")

    print("\nTesting nested capture with callbacks...")
    result2 = test_nested_capture_with_callback()
    print(f"Callback result: {result2}")

    latitude["flush"]()
    print("\nDone! Check Latitude dashboard for context merging verification.")
